docs/development/pigs/pig-009.rst

```diff
@@ -0,0 +1,238 @@
+
+.. code::
+
+    events = vstack([events_src_1, events_src_2, ..., events_src_N])
+    irf_sampler = IRFMapSampler(psf_map, edisp_map)
+
+    events = irf_sampler.sample_psf(events)
+    events = irf_sampler.sample_edisp(events)
+
+Alternatively the IRFs could be applied per model component and be handled in the `MapEventSampler`. However in this case
+the IRFs would not be applied correctly for e.g. very extended sources, because the IRF changes over region of the source.
+If we find the approach with the `IRFMapSampler` to be not performant enough or problematic in memory use, it can be still
+used as a simple fallback solution.
+
+MapDatasetEventSampler?
+=======================
+
+To combine the `MapEventSampler` and `IRFMapSampler` objects we finally propose to introduce a `MapDatasetEventSampler`,
+which executes the sampling from all source components, applies the IRFs, samples from the background model and returns
+a stacked, combined event list.
+
+This is the proposed high level API:
+
+.. code::
+    
+    dataset_sampler = MapDatasetEventSampler(dataset)
+    events = dataset_sampler.sample()
+
+This requires to add `t_min` and `t_max` information to `MapDataset` or maybe a `MapDataset.gti` attribute, defining the
+GTI for the given dataset. In addition it also requires to add a lightcurve model to the `SkyModel`.
+
+More detailed:
+
+.. code::
+
+    events = []
+
+    # sample from source components
+    for evaluator in dataset._evaluators:
+        evaluator.psf = None?
+        evaluator.edisp = None?
+        npred = evaluator.compute_npred()
+
+        sampler = MapEventSampler(npred)
+
+        # set an ID for the source component?
+        
+        events.append(sampler.sample())
+
+    events_stacked = vstack(events)
+
+    # apply IRFs
+    irf_sampler = IRFSampler(events_stacked, dataset.psf_map, dataset.edisp_map)
+    events_reco = irf_sampler.sample()
+
+    # sample from background
+    bkg_sampler = MapEventSampler(dataset.background_model.map)
+    events_bkg = bkg_sampler.sample()
+
+    events_total = vstack([events_reco, event_bkg])
+
+
+Simulating multiple observations
+================================
+The question of simulating mutiple observations from e.g. an `ObservationTable` or `GTI` object as it is needed for
+simulating data for the CTA data challenge is not addressed in this PIG.
+
+
+
+Alternatives
+============
+Instead of introducing a parallel class structure of sampler objects once could introduce `.sample()` methods on the
+existing related classes in Gammapy:
+
+- `events_src = MapEvaluator.sample()` to sample from a source model
+- `events = EdispMap.sample(events=)` to apply the energy dispersion
+- `events = PsfMap.sample(events=)` to apply the PSF
+- `events = MapDataset.sample()` to combine all together.
+
+This has the advantage of a smaller API, but is less flexible.
```

Comment:
I wouldn't call this an alternative really.

If those methods are convenient, one could add them in addition, not instead of the separate sampler class.
Here's such an example, where there's a separate checker class, yet also a check method:
https://github.com/gammapy/gammapy/blob/d6614c845c3c0931da6f5bb3ffd951138fd5e560/gammapy/data/observations.py#L257-L263

Also: you say it's less flexible. Is it? Can you add a sentence or example how it's less flexible?

Reply:
Okay, I included these as additional methods into the `MapEventSampler' class. However, we want to keep the `IRFMapDistributor' class as well, because it allows us to avoid cycles over the number of simulated sources. The other approach can apply the IRF to single source events per time. In this sense, the approach is less flexible.

Reply:
@cdeil Less flexible in the sense, that you can add helper / plotting methods to the `...Sampler` classes and extend the API of these objects if needed. But I agree we could add `.sampler()` methods to our `MapDataset` or `MapEvaluator` class for convenience (it saves the additional import...). So we shouldn't put in the alternative section, but put as an additional proposal, if we want it. I don't have a strong opinion there...